pkg/query/iter.go

```diff
@@ -604,31 +615,20 @@ func (it *dedupSeriesIterator) Next() bool {
 
 	it.useA = ta <= tb
 
-	// For the series we didn't pick, add a penalty twice as high as the delta of the last two
-	// samples to the next seek against it.
-	// This ensures that we don't pick a sample too close, which would increase the overall
-	// sample frequency. It also guards against clock drift and inaccuracies during
-	// timestamp assignment.
-	// If we don't know a delta yet, we pick 5000 as a constant, which is based on the knowledge
-	// that timestamps are in milliseconds and sampling frequencies typically multiple seconds long.
-	const initialPenalty = 5000
-
+	// For the series we didn't pick, add a penalty aligned to the resolution step.
+	// This ensures that we don't miss a sample for the requested resolution when switching iterators and
+	// also not pick too many samples for the current resolution.
+	// Promql has a default look back delta of 5min so if the penalty causes the duration
+	// between 2 samples to be greater then the default look back it causes data gaps.
 	if it.useA {
-		if it.lastT != math.MinInt64 {
-			it.penB = 2 * (ta - it.lastT)
-		} else {
-			it.penB = initialPenalty
-		}
+		it.penB = it.step
```

Comment:
Ok, I am pretty sure step based algo will not work. ): 

Let's take following query: `sum_over_time(my_metric[1h])` for 4d with step 1h. Let's say we have 15s scrape interval of this metric and 2 replicas with rollout in between. So let's say:

`a` replica samples (timestamps after start time): 0, 1, 2, 3,..... 59m58s, 59m59s, 1h, <rollout take 10m>, 1h20m, 1h20m1s, 1h20m2s...
`b` replica samples 1, 2, 3, 4, ..... every second always

Now our algo will choose replica `a` first and add step penalty (`1h`). At the last sample before rollout we will make penalty of B = step (1h), so next iteration we have:

1h20m form a, and 2h from b. This is a gap that could be totally avoided because `b` has that data. This will be highly visible on PromQL result despite the large step, as certain queries cares about each samples  - this has to be reliable ):